File: C++/src/gilded_rose.cpp
#include "GildedRose/gilded_rose.h"

#include <memory>
#include <string>

#include "aged_brie_item.h"
#include "backstage_passes_item.h"
#include "normal_item.h"

namespace
{
const char *const kItemNormal = "Normal Item";
const char *const kItemAgedBrie = "Aged Brie";
const char *const kItemBackstagePasses = "Backstage passes to a TAFKAL80ETC concert";
const char *const kItemSulfuras = "Sulfuras, Hand of Ragnaros";

std::unique_ptr<Item> createItem(const std::string &name, int days_remaining, int quality)
{
    std::unique_ptr<Item> item;

    if (name == kItemNormal) {
        item = std::make_unique<NormalItem>(days_remaining, quality);
    }
    else if (name == kItemAgedBrie) {
        item = std::make_unique<AgedBrieItem>(days_remaining, quality);
    }
    else if (name == kItemSulfuras) {
        item = std::make_unique<Item>(days_remaining, quality);
    }
    else if (name == kItemBackstagePasses) {
        item = std::make_unique<BackstagePassesItem>(days_remaining, quality);
    }
    return item;
}
}  // namespace

GildedRose::GildedRose(const std::string &name, int days_remaining, int quality)
    : _item(createItem(name, days_remaining, quality))
{
}

void GildedRose::tick()
{
    _item->tick();
}

int GildedRose::days_remaining()
{
    return _item->days_remaining();
}

int GildedRose::quality()
{
    return _item->quality();
}

File: Python/gilded_rose.py
class GildedRose:

    def __init__(self, name, days_remaining, quality):
        self.name = name
        self.days_remaining = days_remaining
        self.quality = quality

    def tick(self):
        if self.name != "Aged Brie" and self.name != "Backstage passes to a TAFKAL80ETC concert":
            if self.quality > 0:
                if self.name != "Sulfuras, Hand of Ragnaros":
                    self.quality = self.quality - 1
        else:
            if self.quality < 50:
                self.quality = self.quality + 1
                if self.name == "Backstage passes to a TAFKAL80ETC concert":
                    if self.days_remaining < 11:
                        if self.quality < 50:
                            self.quality = self.quality + 1
                    if self.days_remaining < 6:
                        if self.quality < 50:
                            self.quality = self.quality + 1
        if self.name != "Sulfuras, Hand of Ragnaros":
            self.days_remaining = self.days_remaining - 1
        if self.days_remaining < 0:
            if self.name != "Aged Brie":
                if self.name != "Backstage passes to a TAFKAL80ETC concert":
                    if self.quality > 0:
                        if self.name != "Sulfuras, Hand of Ragnaros":
                            self.quality = self.quality - 1
                else:
                    self.quality = self.quality - self.quality
            else:
                if self.quality < 50:
                    self.quality = self.quality + 1
